### PR TITLE
[Feature] Improved Reachability / Internet-Connection check

### PR DESCRIPTION
**Pull request by SvenTiigi**
## Motivation
Currently the cwa-app-ios is using the [Reachability.swift](https://github.com/ashleymills/Reachability.swift) dependency in order to determine if a valid internet connection is available. 

Reachability.swift does not provide any support for edge cases where the device is connected via a celluar or WiFi connection without internet access for example a [Captive Portal](https://de.wikipedia.org/wiki/Captive_Portal).

Corresponding GitHub issue: https://github.com/ashleymills/Reachability.swift/issues/108

## Solution

Replacing the Reachability.swift Package dependency with the [Connectivity](https://github.com/rwbutler/Connectivity) Swift Package from [Ross Buttler](https://github.com/rwbutler) which enables a robust detection for a valid internet connection.

https://github.com/rwbutler/Connectivity#how-it-works

## Implementation

A dedicated `ReachabilityService` has been added in order to make it more accessible for other components.

`ReachabilityService.swift`
```swift
// MARK: - ReachabilityState

/// The ReachabilityState representing the current state of reachability
/// in terms of if a valid internet connection is available or not
enum ReachabilityState {
	/// Connected. A valid internet connection is available
	case connected
	/// Disconnected. No valid internet connection is available
	case disconnected
}

// MARK: - ReachabilityService

/// The ReachabilityService used to observe the ReachabilityState
/// in order to determine if a Internet connection is available or not
protocol ReachabilityService {
	
	/// Observe ReachabilityState
	/// - Parameters:
	///   - object: The Object to observe on
	///   - observer: The Observer closure
	func observe<Object: AnyObject>(
		on object: Object,
		observer: @escaping (ReachabilityState) -> Void
	)
	
}
```

The function signature uses a generic `AnyObject` in order to automatically dispose the observation when the given object has been deallocated via a `weak` reference.
This a lightweight technique, taken from [John Sundell - Binable values in Swift](https://www.swiftbysundell.com/articles/bindable-values-in-swift/), to dispose the observations as currently the project has no implementation for concepts like `Disposables`, `DisposeBag` or `Cancellables`

`ConnectivityReachabilityService.swift`
```swift
func observe<Object: AnyObject>(
	on object: Object,
	observer: @escaping (ReachabilityState) -> Void
) {
	// Flat map current Connectivity.Status to ReachabilityState
	// and invoke Observer
	self.connectivityStatus
		.flatMap(ReachabilityState.init)
		.flatMap(observer)
	// Append Observer closure by keeping only a weak reference to object
	self.observers.append { [weak object] reachabilityState in
		// Verify weak referenced Object is available
		guard object != nil else {
			// Otherwise return false as object is no longer available
			return false
		}
		// Invoke observer with the given ReachabilityState
		observer(reachabilityState)
		// Return true as the object is still available
		return true
	}
}
```

```swift
func connectivityDidChange(to status: Connectivity.Status) {
	// Update current Connectivity Status
	self.connectivityStatus = status
	// Initialize ReachabilityState from Connectivity Status
	let reachabilityState = ReachabilityState(status)
	// Re-Initialize Observer by invoking each Observer and filter
	// out those which are no longer available
	self.observers = self.observers.filter { $0(reachabilityState) }
}
```

## ENStateHandler

Currently the only component who uses a reachability check is the `ENStateHandler`. Therefore the `ENStateHandler` receives an instance of a `ReachabilityService` via constructor injection and observing it self to `ReachabilityState` updates.

```swift
init(
	_ initialState: ExposureManagerState,
	reachabilityService: ReachabilityService,
	delegate: StateHandlerObserverDelegate
) {
	self.reachabilityService = reachabilityService
	self.delegate = delegate
	currentState = determineCurrentState(from: initialState)
	self.reachabilityService.observe(on: self) { [weak self] reachabilityState in
		self?.internet(reachabilityState == .connected)
	}
}
```

## Tests

The `ENStateTests.swift` file has been updated accordingly to the proposed changes. A dedicated `MockReachabilityService` object has been added which will be constructor injected to the `MockStateHandlerObserverDelegate` to update the `ReachabilityState` inside a test case.

### CODE REVIEW

**Comment by haosap** [comment]
Hi @steins-code ,
Please review and test the PR.
Thanks @SvenTiigi .

**Comment by ChristianKienle** [comment]
@haosap can you check [Connectivity](https://github.com/rwbutler/Connectivity) with the open source team w/r to license etc?

**Comment by ChristianKienle** [comment]
Note to self: Adjust `THIRD-PARTY-NOTICES` to reflect the change once this is merged.

**Comment by haosap** [comment]
> @haosap can you check [Connectivity](https://github.com/rwbutler/Connectivity) with the open source team w/r to license etc?

Yep. Got it.

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift
Comment (steins-code) [inline src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:52]: This line seems not to be called when changing the disconnecting internet. Can you check again?

Comment (SvenTiigi) [reply src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:52]: @steins-code sorry my fault 😄  I've just added a missing invocation for the `setup` function https://github.com/corona-warn-app/cwa-app-ios/pull/53/commits/1ff739bf7b67039690225677f8b5f85446e43992 please try again.

**Review by steins-code — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift
Comment (steins-code) [reply src/xcode/ENA/ENA/Source/Scenes/ENSetting/ENStateHandler.swift:52]: Thanks for your efforts :)

**Comment by steins-code** [comment]
We are waiting for GDPR and license approval before merging this PR...

**Comment by ChristianKienle** [comment]
@steins-code did we get word from our licensing wizards? 🧙‍♂️

**Comment by haosap** [comment]
Hi @SvenTiigi,
Could you please check if the lib is compliant with GDPR? For example, no connection to other servers and no additional storage. Thanks.

**Comment by haosap** [comment]
> @steins-code did we get word from our licensing wizards? 🧙‍♂️

Hi @ChristianKienle ,
The license has been approved. Only need to check the GDPR. I ask @SvenTiigi to do it.
Thanks.

**Comment by SvenTiigi** [comment]
Hey @haosap

I've checked out the code of Connectivity in Version [3.3.4](https://github.com/rwbutler/Connectivity/releases/tag/3.3.4). 
Here are my inspection results:

- Connectivity does not perform any kind of additional storage. No UserDefaults, no CloudKit or NSUbiquitousKeyValueStore, no local Storage such as accessing the file-system or CoreData.
- No information about the device or the user will be accessed or gathered.

What Connectivity does is to connect to special servers from Apple in order to check if the device is currently inside a Captive Portal.

`Connectivity.swift L:334`
```
/// Set of connectivity URLs used by default if none are otherwise specified.
static func defaultConnectivityURLs(shouldUseHTTPS: Bool) -> [URL] {
    var result: [URL] = []
    let connectivityURLs: [String] = shouldUseHTTPS
        ? ["https://www.apple.com/library/test/success.html",
           "https://captive.apple.com/hotspot-detect.html"] // Replace with custom URLs
        : ["http://www.apple.com/library/test/success.html",
           "http://apple.com/library/test/success.html",
           "http://www.appleiphonecell.com/library/test/success.html",
           "http://www.itools.info/library/test/success.html",
           "http://www.ibook.info/library/test/success.html",
           "http://www.airport.us/library/test/success.html",
           "http://www.thinkdifferent.us/library/test/success.html",
           "http://captive.apple.com/hotspot-detect.html"]
    for connectivityURLStr in connectivityURLs {
        if let connectivityURL = URL(string: connectivityURLStr) {
            result.append(connectivityURL)
        }
    }
    return result
}
```

If the `NSAllowsArbitraryLoads` in the Info.plist is not present or set to false, Connectivity will only connect to `https` server from Apple https://www.apple.com/library/test/success.html and https://captive.apple.com/hotspot-detect.html in order to retrieve the following HTML content:

```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
	<TITLE>Success</TITLE>
</HEAD>
<BODY>
	Success
</BODY>
</HTML>
```
Read more: https://github.com/rwbutler/Connectivity#how-it-works

**Comment by haosap** [comment]
> Hey @haosap
> 
> I've checked out the code of Connectivity in Version [3.3.4](https://github.com/rwbutler/Connectivity/releases/tag/3.3.4).
> Here are my inspection results:
> 
> * Connectivity does not perform any kind of additional storage. No UserDefaults, no CloudKit or NSUbiquitousKeyValueStore, no local Storage such as accessing the file-system or CoreData.
> * No information about the device or the user will be accessed or gathered.
> 
> What Connectivity does is to connect to special servers from Apple in order to check if the device is currently inside a Captive Portal.
> 
> `Connectivity.swift L:334`
> 
> ```
> /// Set of connectivity URLs used by default if none are otherwise specified.
> static func defaultConnectivityURLs(shouldUseHTTPS: Bool) -> [URL] {
>     var result: [URL] = []
>     let connectivityURLs: [String] = shouldUseHTTPS
>         ? ["https://www.apple.com/library/test/success.html",
>            "https://captive.apple.com/hotspot-detect.html"] // Replace with custom URLs
>         : ["http://www.apple.com/library/test/success.html",
>            "http://apple.com/library/test/success.html",
>            "http://www.appleiphonecell.com/library/test/success.html",
>            "http://www.itools.info/library/test/success.html",
>            "http://www.ibook.info/library/test/success.html",
>            "http://www.airport.us/library/test/success.html",
>            "http://www.thinkdifferent.us/library/test/success.html",
>            "http://captive.apple.com/hotspot-detect.html"]
>     for connectivityURLStr in connectivityURLs {
>         if let connectivityURL = URL(string: connectivityURLStr) {
>             result.append(connectivityURL)
>         }
>     }
>     return result
> }
> ```
> 
> If the `NSAllowsArbitraryLoads` in the Info.plist is not present or set to false, Connectivity will only connect to `https` server from Apple https://www.apple.com/library/test/success.html and https://captive.apple.com/hotspot-detect.html in order to retrieve the following HTML content:
> 
> ```
> <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
> <HTML>
> <HEAD>
> 	<TITLE>Success</TITLE>
> </HEAD>
> <BODY>
> 	Success
> </BODY>
> </HTML>
> ```
> 
> Read more: https://github.com/rwbutler/Connectivity#how-it-works

Hmm... , some concern here: 
http://www.ibook.info/library/test/success.html",
http://www.airport.us/library/test/success.html",
http://www.thinkdifferent.us/library/test/success.html"

**Comment by SvenTiigi** [comment]
> Hmm... , some concern here:
> http://www.ibook.info/library/test/success.html",
> http://www.airport.us/library/test/success.html",
> http://www.thinkdifferent.us/library/test/success.html"

When initializing a Connectivity instance you can pass the parameter `shouldUseHTTPS` which is in default set to `true`.

```swift
public init(shouldUseHTTPS: Bool = true) {
    type(of: self).isHTTPSOnly = shouldUseHTTPS
    self.reachability = Reachability.forInternetConnection()
}
```

This means my previous post was not correct. Connectivity is choosing the URLs based on the static property `isHTTPSOnly` which is in default set to true. The only part where the `NSAllowsArbitraryLoads` comes in play is when you try to disable `isHTTPSOnly`. Connectivity will then check if it is even possible to not use HTTPS and if not it will reset the static property to `true`.

```swift
/// Whether or not only HTTPS URLs should be used to check connectivity
public static var isHTTPSOnly: Bool = true {
    didSet {
        // Only set true if `allow arbitrary loads` is set
        guard let bundleInfo = Bundle.main.infoDictionary,
            let appTransportSecurity = bundleInfo["NSAppTransportSecurity"] as? [String: Any],
            let allowsArbitraryLoads = appTransportSecurity["NSAllowsArbitraryLoads"] as? Bool,
            allowsArbitraryLoads else {
            isHTTPSOnly = true
            return
        }
    }
}
```

So the `http://` URLs will never been used unless the cwa-app-ios contains the `NSAllowsArbitraryLoads` key with a value of `true` and the Connectivity static property `isHTTPSOnly` is manually changed to `false`.

This PR never sets the `isHTTPSOnly` static property to false so in conclusion only the `https://` URLs will be used.

**Comment by haosap** [comment]
Hi @SvenTiigi,
Great explanation! After the GDPR clarified, we can merge it.
Please also solve the conflict. Thanks.

**Comment by SvenTiigi** [comment]
Done ✅
Merged upstream branch and resolved merged conflicts in `ENStateTests.swift` file

**Review by haosap — APPROVED** [review]
good job!

**Review by ChristianKienle — CHANGES_REQUESTED** [review]
I am sorry but I have to veto this for now. We experience build issues at the moment that are caused by third party libraries. We want to evaluate any PR that introduces new dependencies from now on w/r to potential build issues. We are in a critical phase and thus have to be careful. We will definitely look at this PR again tomorrow so I will only "veto" this over night so that we don't end up with something we can't deploy.

Hope this is okay.

**Comment by ChristianKienle** [comment]
@SvenTiigi We have discussed this and we would love to merge this. Could you resolve the merge conflicts? Then we could merge it straight away...

**Comment by SvenTiigi** [comment]
Merged ✅

**Review by ChristianKienle — APPROVED** [review]
Thank you very much. <3

**Pull request merged**
